Deprecate the methods of `default.qubit`

mudit2812 opened the pull request:
**Description of the Change:**
* Added deprecation warnings to user-facing `DefaultQubit` methods that aren't present in `DefaultQubit2` to recommend using `DefaultQubit2` instead.
* Updated `QubitDevice` `execute` and `batch_execute`, as well as `Device.check_validity` to mute the warning during normal execution so that warnings are only raised when users directly call the deprecated methods.

mudit2812 commented:
[sc-37948]

trbromley commented on pennylane/devices/default_qubit.py:
```diff
@@ -44,6 +45,14 @@
 
+def deprecation_warning(func_name):
+    warnings.warn(
+        f"DefaultQubit.{func_name} is deprecated. Use qml.devices.experimental.DefaultQubit2 "
+        "instead.",
```
I'm not sure if we're ready to recommend DefaultQubit2 as a replacement. @albi3ro?
Maybe something like:
> Directly calling the methods of default.qubit is not recommended and should be avoided where possible. The available methods of default.qubit will be updated soon to follow a new API described here: https://docs.pennylane.ai/en/stable/code/api/pennylane.devices.experimental.DefaultQubit2.html

mudit2812 replied:
I've added filters to `QubitDevice.execute`, `QubitDevice.batch_execute`, `Device.check_validity`, so the warnings won't show up in user facing workflows for executing qnodes/tapes. In all other cases, the warning will show up. So in most cases, the warning will only show up when a user directly accesses the functions being deprecated.

albi3ro commented on doc/development/deprecations.rst:
```diff
@@ -51,6 +51,11 @@ Pending deprecations
         some_qfunc(params)
         return qml.expval(Hamiltonian)
 
+* The public methods of `DefaultQubit` that aren't present in `DefaultQubit2` now raise a deprecation
+  warning if called directly by users. The warning is not raised in normal workflows.
+
+  - Deprecated in v0.31
+
```
I wonder if its worth adding some more context here, along the lines of 

```
We are switching to a new device interface. In this process, simulation implementation details are abstracted away from the device class itself and provided by composition, rather than inheritance. Therefore, many public and private methods from `DefaultQubit` will no longer exist, though its behaviour in a workflow will remain the same. See :class:`pennylane.devices.experimental.Device` for more information on what the new interface will look like.

To directly work with the components of a Python state vector simulation, please consult the `devices.qubit` submodule.
```

I wonder if we should add a table of contents for `devices.qubit` on the `devices` "Internals" page.

timmysilv reviewed (approved):
nice!

trbromley commented on pennylane/devices/default_qubit.py:
```diff
@@ -204,6 +204,17 @@ def __init__(
 
     @property
     def stopping_condition(self):
+        """.BooleanFn: Returns the stopping condition for the device. The returned
```
Is the output actually of `BooleanFn` type?

mudit2812 replied:
Yep it is.

trbromley commented on pennylane/devices/default_qubit.py:
```diff
@@ -204,6 +204,17 @@ def __init__(
+        .. warning::
+
+            Directly using this property is not recommended and should be avoided where possible.
+            The available methods of DefaultQubit will be updated soon to follow a new API described
```
```suggestion
            The API of DefaultQubit will be updated soon to follow a new interface described
```

trbromley replied:
(same for further down too)

trbromley commented on pennylane/devices/default_qubit.py:
```diff
@@ -249,6 +260,19 @@ def _get_batch_size(self, tensor, expected_shape, expected_size):
 
     # pylint: disable=arguments-differ
     def apply(self, operations, rotations=None, **kwargs):
+        """Apply a sequence of operations and rotations to the device.
+
+        .. warning::
+
+            Directly calling this method is not recommended and should be avoided where possible.
+            The available methods of DefaultQubit will be updated soon to follow a new API described
+            `here <https://docs.pennylane.ai/en/stable/code/api/pennylane.devices.experimental.DefaultQubit2.html>`_.
+
+        Args:
+            operations (List[.Operation]): List of operations to apply
+            rotations (List[.Operation]): List of rotations to apply to rotate state to measurement basis
+        """
```
These are the docstrings already available for [apply](https://docs.pennylane.ai/en/stable/code/api/pennylane.devices.default_qubit.DefaultQubit.html#pennylane.devices.default_qubit.DefaultQubit.apply):
![image](https://github.com/PennyLaneAI/pennylane/assets/49409390/ee77a1ed-6430-46e9-9432-6fdf05c98cd1)

Have you shortened them consciously to dissuade users from using this method?

trbromley reviewed (commented):
Thanks @mudit2812!

One overarching question is _why these methods/attributes in particular_? [DefaultQubit](https://docs.pennylane.ai/en/stable/code/api/pennylane.devices.default_qubit.DefaultQubit.html) has other methods and attributes unavailable in [DefaultQubit2](https://docs.pennylane.ai/en/stable/code/api/pennylane.devices.experimental.DefaultQubit2.html), e.g., `author`. Do we think the methods and attributes we've added warning messages to here are the main ones users will be affected by?

mudit2812 commented on pennylane/devices/default_qubit.py:
```diff
@@ -249,6 +260,19 @@ def _get_batch_size(self, tensor, expected_shape, expected_size):
+        .. warning::
+
+            Directly calling this method is not recommended and should be avoided where possible.
+            The available methods of DefaultQubit will be updated soon to follow a new API described
```
```suggestion
            The API of DefaultQubit will be updated soon to follow a new interface described
```

mudit2812 commented on pennylane/devices/default_qubit.py:
```diff
@@ -535,6 +559,12 @@ def expval(self, observable, shot_range=None, bin_size=None):
+        .. warning::
+
+            Directly calling this method is not recommended and should be avoided where possible.
+            The available methods of DefaultQubit will be updated soon to follow a new API described
```
```suggestion
            The API of DefaultQubit will be updated soon to follow a new interface described
```

mudit2812 commented on pennylane/devices/default_qubit.py:
```diff
@@ -683,6 +714,15 @@ def _create_basis_state(self, index):
+        .. warning::
+
+            Directly requesting this property is not recommended and should be avoided where possible.
+            The available methods of DefaultQubit will be updated soon to follow a new API described
```
```suggestion
            The API of DefaultQubit will be updated soon to follow a new interface described
```

mudit2812 commented on pennylane/devices/default_qubit.py:
```diff
@@ -900,14 +940,37 @@ def _apply_diagonal_unitary(self, state, phases, wires):
+        .. warning::
+
+            Directly calling this method is not recommended and should be avoided where possible.
+            The available methods of DefaultQubit will be updated soon to follow a new API described
```
```suggestion
            The API of DefaultQubit will be updated soon to follow a new interface described
```

mudit2812 commented on pennylane/devices/default_qubit.py:
```diff
@@ -900,14 +940,37 @@ def _apply_diagonal_unitary(self, state, phases, wires):
+        .. warning::
+
+            Directly calling this method is not recommended and should be avoided where possible.
+            The available methods of DefaultQubit will be updated soon to follow a new API described
```
```suggestion
            The API of DefaultQubit will be updated soon to follow a new interface described
```

mudit2812 commented:
> One overarching question is why these methods/attributes in particular? [DefaultQubit](https://docs.pennylane.ai/en/stable/code/api/pennylane.devices.default_qubit.DefaultQubit.html) has other methods and attributes unavailable in [DefaultQubit2](https://docs.pennylane.ai/en/stable/code/api/pennylane.devices.experimental.DefaultQubit2.html), e.g., author. Do we think the methods and attributes we've added warning messages to here are the main ones users will be affected by?

I added warnings for the methods I would imagine getting used the most often. As for the attributes like `author`, I have changed them into properties so that the warnings can render in the docs for them.

trbromley commented on pennylane/devices/default_qubit.py:
```diff
@@ -86,11 +86,57 @@ class DefaultQubit(QubitDevice):
+        .. warning::
+
+            Directly using this property is not recommended and should be avoided where possible.
+            The API of DefaultQubit will be updated soon to follow a new interface described
```
```suggestion
            The API of ``DefaultQubit`` will be updated soon to follow a new interface described
```
Though this is in many lines below :sweat_smile:

albi3ro reviewed (approved):
👍

trbromley commented on doc/development/plugins.rst:
```diff
@@ -58,6 +58,12 @@ and subclassing it:
 
+.. warning::
+
+    The API of :class:`~.DefaultQubit` will be updated soon to follow a new interface
+    defined by the experimental :class:`pennylane.devices.experimental.Device` class. To view the
+    exact changes coming to :class:`~.DefaultQubit`, see :class:`pennylane.devices.experimental.DefaultQubit2`.
```
```suggestion
    The API of PennyLane devices will be updated soon to follow a new interface defined by
    the experimental :class:`pennylane.devices.experimental.Device` class. This guide describes
    how to create a device with the current :class:`pennylane.Device` and
    :class:`pennylane.QubitDevice` base classes, and will be updated as we switch to the new API.
    In the meantime, please reach out to the PennyLane team if you would like help with building
    a plugin, either by creating an [issue](https://github.com/PennyLaneAI/pennylane/issues) or
    by posting in our [discussion forum](https://discuss.pennylane.ai/).
```
For this particular page, we need to be a bit more general than for default.qubit.

trbromley commented on doc/releases/changelog-dev.md:
```diff
@@ -333,6 +333,12 @@
 
+* The public methods of `DefaultQubit` are pending changes to
+  follow the new device API, as used in `DefaultQubit2`.
+  The docstrings of the affected methods in `DefaultQubit` have been
+  updated to reflect this.
```
```suggestion
  Warnings have been added to the docstrings to reflect this.
```

trbromley commented on doc/development/deprecations.rst:
```diff
@@ -98,6 +98,26 @@ Pending deprecations
+  for more information on what the new interface will look like and be prepared
+  to make updates in a coming release. If you have any feedback on these
+  changes, please create an
+  [issue](https://github.com/PennyLaneAI/pennylane/issues) or post in our
```
```suggestion
  `issue <https://github.com/PennyLaneAI/pennylane/issues>`__ or post in our
```

trbromley commented on doc/development/deprecations.rst:
```diff
@@ -98,6 +98,26 @@ Pending deprecations
+  to make updates in a coming release. If you have any feedback on these
+  changes, please create an
+  [issue](https://github.com/PennyLaneAI/pennylane/issues) or post in our
+  [discussion forum](https://discuss.pennylane.ai/).
```
```suggestion
  `discussion forum <https://discuss.pennylane.ai/>`__.
```

trbromley commented on pennylane/devices/default_mixed.py:
```diff
@@ -48,6 +48,11 @@
+
+        The API of ``DefaultMixed`` will be updated soon to follow a new device interface described
+        in :class:`pennylane.devices.experimental.Device`.
+
```
```suggestion
        This change will not alter device behaviour for most workflows, but may have implications for
        plugin developers and users who directly interact with device methods.
        Please consult :class:`pennylane.devices.experimental.Device` and the implementation in
        :class:`pennylane.devices.experimental.DefaultQubit2`
        for more information on what the new interface will look like and be prepared
        to make updates in a coming release. If you have any feedback on these
        changes, please create an
        `issue <https://github.com/PennyLaneAI/pennylane/issues>`__ or post in our
        `discussion forum <https://discuss.pennylane.ai/>`__.

```

trbromley commented on pennylane/devices/default_qubit.py:
```diff
@@ -77,6 +77,12 @@ def _get_slice(index, axis, num_axes):
+        The API of ``DefaultQubit`` will be updated soon to follow a new device interface described
+        in :class:`pennylane.devices.experimental.Device`. To view the exact changes coming to
+        ``DefaultQubit``, see :class:`pennylane.devices.experimental.DefaultQubit2`.
+
```
```suggestion
        This change will not alter device behaviour for most workflows, but may have implications for
        plugin developers and users who directly interact with device methods.
        Please consult :class:`pennylane.devices.experimental.Device` and the implementation in
        :class:`pennylane.devices.experimental.DefaultQubit2`
        for more information on what the new interface will look like and be prepared
        to make updates in a coming release. If you have any feedback on these
        changes, please create an
        `issue <https://github.com/PennyLaneAI/pennylane/issues>`__ or post in our
        `discussion forum <https://discuss.pennylane.ai/>`__.

```

trbromley commented on pennylane/devices/default_qutrit.py:
```diff
@@ -36,6 +36,11 @@
+
+        The API of ``DefaultQutrit`` will be updated soon to follow a new device interface described
+        in :class:`pennylane.devices.experimental.Device`.
+
```
```suggestion
        This change will not alter device behaviour for most workflows, but may have implications for
        plugin developers and users who directly interact with device methods.
        Please consult :class:`pennylane.devices.experimental.Device` and the implementation in
        :class:`pennylane.devices.experimental.DefaultQubit2`
        for more information on what the new interface will look like and be prepared
        to make updates in a coming release. If you have any feedback on these
        changes, please create an
        `issue <https://github.com/PennyLaneAI/pennylane/issues>`__ or post in our
        `discussion forum <https://discuss.pennylane.ai/>`__.

```

trbromley commented on pennylane/devices/null_qubit.py:
```diff
@@ -30,6 +30,11 @@ class NullQubit(QubitDevice):
+
+        The API of ``NullQubit`` will be updated soon to follow a new device interface described
+        in :class:`pennylane.devices.experimental.Device`.
+
```
```suggestion
        This change will not alter device behaviour for most workflows, but may have implications for
        plugin developers and users who directly interact with device methods.
        Please consult :class:`pennylane.devices.experimental.Device` and the implementation in
        :class:`pennylane.devices.experimental.DefaultQubit2`
        for more information on what the new interface will look like and be prepared
        to make updates in a coming release. If you have any feedback on these
        changes, please create an
        `issue <https://github.com/PennyLaneAI/pennylane/issues>`__ or post in our
        `discussion forum <https://discuss.pennylane.ai/>`__.

```

trbromley reviewed (commented):
Thanks @mudit2812, overall looks good I've just left suggestions to tweak the warnings.
